### ISSUE TITLE
cypress not able to stop to load xhr with window object

### ISSUE THREAD

**Issue by MehranShafqat** [cypress-io/cypress#7083]
Well I try to stop the  load the xhr and continue my test without page load time out error
but  not working and not even  stop the xhr loading.  Please guide if im using in correct way or it is actual bug.

cy.window((win) => {
            win.stop()
          })

cy version 4.4.0
OS windows 10

**Comment by MehranShafqat** [comment]
@jennifer-shehane  please have a look

**Comment by jennifer-shehane** [comment]
`blacklistHosts` only accepts *hosts*, not fully qualified urls: https://developer.mozilla.org/en-US/docs/Web/API/URL/host

It doesn't appear that the `window.stop` application will work here because once Cypress sees a page load, it waits for the page to load before moving to the next Cypress step. 

I can't think of a way to stop loading of a specific XHR in Cypress currently.

**Comment by jennifer-shehane** [comment]
I found a way to issue `win.stop()` to a loading page and have added an example below to demonstrate. Basically, you need to listen to the [`window:before:load` event](https://on.cypress.io/catalog-of-events#App-Events) in order to stop the window from loading how you want. 

In my example, we visit `docs.cypress.io` and click a link that routes us to `www.cypress.io`. We can listen in the `window:before:load` for the exact href we want to not load and prevent the load there. 

```js
it('test', () => {
  // set up the listener before the action that will cause reroute
  cy.on('window:before:load', (win) => {
    // just log the win.location.href for convenience
    console.log('WINDOW BEFORE LOAD', win.location.href)

    // if we're trying to load the page we want to stop, win.stop()
    if (win.location.href === 'https://www.cypress.io/') {
      win.stop()
    }
  })

  cy.visit('https://docs.cypress.io')
  cy.get('#logo').click()
})
```

If you want the listener to be global (across all tests), make sure to use `Cypress.on` for the listener instead of `cy.on`.

**HOWEVER**, this does not prevent Cypress from listening to the page load. Cypress still waits for the page to load - so since you've prevented the page from loading, the test will now fail after the default 60 second timeout. This doesn't seem like a great solution to me. 

<img width="922" alt="Screen Shot 2020-04-24 at 4 32 16 PM" src="https://user-images.githubusercontent.com/1271364/80200806-32b9e680-8649-11ea-875c-7b1d89e0a5fc.png">

**Comment by norvinino** [comment]
Clicking this download link works fine, but I can't stop cypress from waiting for a nonexistant page to load, which prevents the rest of the test from running.
![image](https://user-images.githubusercontent.com/71896864/94316339-c3bf2180-ff49-11ea-8851-6766808dcb6a.png)

**Comment by norvinino** [comment]
@ jennifer-shehane

**Comment by anjalichhabra29** [comment]
> Clicking this download link works fine, but I can't stop cypress from waiting for a nonexistant page to load, which prevents the rest of the test from running.
> ![image](https://user-images.githubusercontent.com/71896864/94316339-c3bf2180-ff49-11ea-8851-6766808dcb6a.png)

Any solution on this ?

**Comment by tuan231195** [comment]
👍 , this would be a really useful feature

**Comment by jennifer-shehane** [comment]
@norvinino No, there is no solution for this which is why this issue is open.

**Comment by elliottyates** [comment]
In our case, a button in our application fires an XHR to get a temporary URL for a download, then uses `window.location.assign()` with `response-content-disposition=attachment` in order to download a file. I think it would be possible to check for the file on disk based on the cypress examples, but it appears that the fact cypress waits for the "new page" to load prevents those assertions from running. (`window.location.assign()` cannot be `stub()`bed, unfortunately, as that would be all we'd need, but that is another matter!)

I think this use case is related to a number of open and closed issues, but wasn't sure where to include it.

**Comment by elliottyates** [comment]
The workaround I'm using, I should mention, is the `blockHosts` feature, which will stop the load and allow assertions to run. (I suppose that might not work if the file is not being downloaded from a discernibly different host than the app is being served from.)

**Comment by ZVince9** [comment]
is there any ETA on this?

Thank you!

**Comment by jimpriest** [comment]
I'm running into similar issues.  Was playing around with the new download feature in 6.3.0 but that functionality seems to be maybe targeted at simply downloading a file from a link.   I'm dynamically generating a file (via ColdFusion) and serving it through the browser. 

So in 6.3.0 I submit my form, the file is downloaded to cypress/downloads (yay)  but Cypress fails while waiting for a page load which never happens (boo) 

This is also similar to #8619

**Comment by suchitrachavan** [comment]
Hi Team I am also facing similar issues 
When i click on a button
one unexpected popup (window,URL) opens up
then cypress fails

**Comment by suchitrachavan** [comment]
![image](https://user-images.githubusercontent.com/61357214/105276743-12836d80-5b70-11eb-9e1e-190740a6f747.png)

**Comment by jennifer-shehane** [comment]
@jimpriest Can you [open a new issue](https://github.com/cypress-io/cypress/issues/new/choose) detailing how this never loads the page when testing file download? We'd like to look into this. 

As mentioned, 6.3.0 does offer testing file downloads in case this was anyone elses blocker in this issue. Please update and remove any workaround code  you had for testing downloads. If you encounter a new issue - please open a new issue. Thanks.

**Comment by suchitrachavan** [comment]
@jennifer-shehane  Do you have any workaround as i tried all the workarounds to handle window and page alerts ...I am using electron browser and can't use chrome & firefox due to security concerns in my organization .Please help

**Comment by aminao-toptal** [comment]
Hi! This is huge blocker for my test suite. Did anyone looked into this problem?

**Comment by jennifer-shehane** [comment]
There is no workaround. You need to avoid the action that is causing the redirect at the moment until this feature is added to Cypress.

**Comment by jimpriest** [comment]
@jennifer-shehane  I created #14789

**Comment by szanelato** [comment]
@jennifer-shehane I'm facing a similar issue when testing a login action. Can you take a look here please? https://github.com/cypress-io/cypress/discussions/15849

Thanks.

**Comment by anastasiiaols** [comment]
Hi, I face the same issue when downloading files. The download itself works fine, but then Cypress waits for some non-existant page load and eventually fails before it can verify the downloaded file. I wasn't able to find any workarounds for it so would much appreciate if the option to stop the page load will be added to Cypress. Is there any ETA for it? Thanks!
<img width="1015" alt="Screen Shot 2021-06-08 at 10 40 28 AM" src="https://user-images.githubusercontent.com/17346034/121233196-3d344100-c847-11eb-85af-d26392c4a1b1.png">

**Comment by manasaj111** [comment]
Hi, I am also facing the same issue when downloading files. Please suggest any work around for this.

**Comment by manasaj111** [comment]
I found a work around here - #14857

cy.window().document().then(function (doc) {
  doc.addEventListener('click', () => {
    setTimeout(function () { doc.location.reload() }, 5000)
  })
  cy.get('[ng-click="vm.export()"]').click()
})

Thanks to Sarja 👍

**Comment by melibe23** [comment]
I am facing this issue with my Google Analytics events coverage. On my tests I stubbed Google Analytics events on every action, these events are fired before the new page loads, so I don't **need** to wait for this load in order to assert, yet Cypress forced the test to wait making the time execution last longer than it could. Is there any workaround that doesn't involved block hosting? Thank you.

**Comment by chaitanyaraju2** [comment]
I am facing the same issue in cypress 
![image](https://user-images.githubusercontent.com/43433069/154855225-d58a75e0-536d-4644-b7f4-1c1a86d20fde.png)
But in selenium, I am able to complete my tests.
waiting for the solution/fix in cypress.

**Comment by samtsai** [comment]
In my case I was able to manually trigger the Page Load event:
https://github.com/cypress-io/cypress/issues/14857#issuecomment-1044633756

**Comment by sbarba** [comment]
Complete blocker for me. We need a clean way to tell Cypress not to _wait for a non-existent page_.

**Comment by astraljames** [comment]
Same here. We can't test any of our downloads as we use an iframe pointing towards an AWS URL. The page load never happens and we can't move forward with testing one of our most critical functions.

**Comment by JoaqPareja** [comment]
Heyy just tried this just throw me the following error:
![9](https://user-images.githubusercontent.com/80791761/200952952-f63d49cd-5b0b-447e-886a-ab805531e742.JPG)

**Comment by tina-zeng** [comment]
I've fixed a similar issue. My page redirects with the parameter ?target=http://xxxx.com/xxx, Just replacing the 'http:' with 'https:' can resolve this issue.

**Comment by MadiyarTastanbekov** [comment]
> I've fixed a similar issue. My page redirects with the parameter ?target=http://xxxx.com/xxx, Just replacing the 'http:' with 'https:' can resolve this issue.
hi how did you solve the problem?

**Comment by MadiyarTastanbekov** [comment]
> Эй, только что попробовал, просто выдал мне следующую ошибку: ![9](https://user-images.githubusercontent.com/80791761/200952952-f63d49cd-5b0b-447e-886a-ab805531e742.JPG)

hi how did you solve the problem

**Comment by MadiyarTastanbekov** [comment]
> Полный блокировщик для меня. Нам нужен чистый способ сказать Cypress не _ждать несуществующей страницы_ .

hi how did you solve the problem

**Comment by tina-zeng** [comment]
> > Эй, только что попробовал, просто выдал мне следующую ошибку: ![9](https://user-images.githubusercontent.com/80791761/200952952-f63d49cd-5b0b-447e-886a-ab805531e742.JPG)
> 
> hi how did you solve the problem

You can try to force visit the (new url) from your screenshot with HTTPS, no need to add scripts to reload.

**Comment by violetflash** [comment]
> Heyy just tried this just throw me the following error: ![9](https://user-images.githubusercontent.com/80791761/200952952-f63d49cd-5b0b-447e-886a-ab805531e742.JPG)


```javascript
cy.intercept('GET', downloadUrl).as('downloadFile');
cy.doIfDataCySelectorExist(selector, () => {
  cy.window().then(function (win) {
    win.addEventListener('click', () => {
      setTimeout(function () {
        win.location.reload()
      }, 5000)
    })
    cy.getByDataCySelector(selector).click();
    cy.wait('@downloadFile').its('response.statusCode').should('eq', 200);
  })
})
```

**Comment by sbarba** [comment]
Nearly 3 years now. Should we conclude there's no interest in fixing this?

**Comment by boonier** [comment]
Tis a real kicker this. Download file served from a tokenised url here. File downloads absolutely fine. Cypress doesn't know what to do. Triggering the download by setting the window.location `window.location = url;`

**Comment by JudeaDumont** [comment]
I got passed it by just catching the error via pasting this before the submit:

    cy.once("fail", (err) =>
    {
        return false;
    });

**Comment by scramlo** [comment]
> I got passed it by just catching the error via pasting this before the submit

LOL! That is one way to do it. This makes tests unfathomably slow as they wait for Cypress to retry for 30 seconds for each download assertion.

**Comment by tushki225** [comment]
I am also facing this issue and assuming its still not fixed.
<img width="965" alt="Capture" src="https://github.com/cypress-io/cypress/assets/18591785/d8b19eb4-6faf-4a45-a3a8-9c072789918b">

**Comment by stevessp** [comment]
Yep. It's been *years* and multiple releases and they've shown zero interest in fixing it.

**Comment by JudeaDumont** [comment]
> > I got passed it by just catching the error via pasting this before the submit
> 
> LOL! That is one way to do it. This makes tests unfathomably slow as they wait for Cypress to retry for 30 seconds for each download assertion.

That probably depends on timeout settings and what you are waiting for (might not be a download, like it wasn't for me). But yeah it's hilarious.

One thing you might also try is hitting the endpoint through jQuery instead of cypress. But you might find that equally hilarious for some reason specific to your own random use case. idk.

**Comment by adnanerlansyah403** [comment]
Are there any updates for this issue ?

**Comment by adnanerlansyah403** [comment]
I'm so frustrating with this issue

**Comment by aditya3151** [comment]
![Screenshot 2024-01-23 164931](https://github.com/cypress-io/cypress/assets/18242663/17e891f1-7282-4a17-8b9c-02fbe7a9c6a0)
I'm also facing similar issue, when ever I tried clicking on the element the page is loading and rest of test commands are failing

**Comment by araujosnathan** [comment]
I am facing the same issue here!
I need to continue with the tests but it is failing because the Cypress waits for a loaded page when it does not exist.

**Comment by ErcPotato** [comment]
any solution for this?

**Comment by sbarba** [comment]
They do not care I guess. Dealbreaker for me.

**Comment by GULNURKIRCAER** [comment]
Is there still no solution for this page load event issue? This is a big problem that affects the execution of our tests. Please take care of it @cypress-app-bot

**Comment by poleselfg** [comment]
The error still persists

**Comment by GULNURKIRCAER** [comment]
yes, the error still persists for me too.
I'm so frustrating with this issue
I'm losing a lot of time in tests because of this event. 
@cypress-app-bot could you help for it
